Adding SupportedTargetFrameworkAlias itemgroup

sfoslund opened the pull request.

sfoslund committed:
Adding SupportedTargetFrameworkAlias itemgroup for project system consumption

davkean commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateSupportedTargetFrameworkAlias.cs:
```diff
@@ -0,0 +1,50 @@
+            foreach (var tfm in SupportedTargetFramework)
+            {
+                var targetFrameworkAlias = NuGetFramework.ParseComponents(tfm.ItemSpec, TargetPlatformMoniker).GetShortFolderName();
+                var newItem = new TaskItem(tfm);
```
This causes it to inherit the metadata?

sfoslund replied:
Yes, this copies the item including the metadata.

sfoslund committed:
Removing build hook from GenerateSupportedTargetFrameworkAlias target

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateSupportedTargetFrameworkAlias.cs:
```diff
@@ -0,0 +1,49 @@
+        public ITaskItem[] SupportedTargetFramework { get; set; }
+
+        [Required]
+        public string TargetFrameworkMoniker { get; set; }
+
+        public string TargetPlatformMoniker { get; set; }
```
I'm not following how this design is supposed to work.  The list of items in the dropdown shouldn't depend on what the current target framework is, so I don't think this task should be taking the current target framework and platform as parameters.

I think it should basically take UseWpf and UseWindowsForms as arguments, and if either of them is true, it should remove any supported frameworks below .NET Core 3.0, and add "-windows" to any of them that are 5.0 or higher.

sfoslund replied:
Currently the target framework drop down is bound to the `TargetFrameworkMoniker`, so part of this change is changing it to bind to the `TargetFramework` instead (see [this line](https://github.com/dotnet/sdk/pull/13545/files/1d6963e324995cee06ccdae90325a54c146e8eb9#diff-11e7b7431abbd0596f030eefc10620e9R42)). 

For generality, I'm constructing the target framework from both the `TargetFrameworkMoniker` and `TargetPlatformMoniker` values. This requires the special case [here](https://github.com/dotnet/sdk/pull/13545/files/1d6963e324995cee06ccdae90325a54c146e8eb9#diff-11e7b7431abbd0596f030eefc10620e9R33) because the `TargetPlatformMonkier` defaulted to windows 7.0 before 5.0, so we want to remove that unless we really are targeting windows (that is, UseWPF or UseWindowsForms are set).

dsplaisted replied:
What are the valid `TargetFramework` values for a Windows Forms or WPF project?  It should be this:

- `netcoreapp3.0`
- `netcoreapp3.1`
- `net5.0-windows`

That list doesn't depend on what the current target framework or platform is.  So I would write the code something like this:

```c#
foreach (var tfm in SupportedTargetFramework)
{
    var targetPlatform = "";
    var targetFramework = NuGetFramework.Parse(tfm.ItemSpec);
    if (UseWindowsForms || UseWpf)
    {
        if (targetFramework.Version.Major >= 5)
        {
            targetFramework.Platform = "Windows";
        }
        if (targetFramework.Version.Major < 3)
        {
            //  Windows Forms / WPF not supported below .NET Core 3.0
            continue;
        }
    }
    var targetFrameworkAlias = targetFramework.GetShortFolderName();
    var displayName = string.IsNullOrWhiteSpace(tfm.GetMetadata("DisplayName"))? targetFrameworkAlias : tfm.GetMetadata("DisplayName");
    convertedTfms.Add(new TaskItem(targetFrameworkAlias, new Dictionary<string, string>() { { "DisplayName", displayName } }));
}
```

(This code doesn't account for projects that target something besides .NETCoreApp, so that would also need to be taken into consideration).

sfoslund replied:
My motivation for using the target platform is generality in the future so that we're minimizing the special casing for windows. This way the code works for target platforms besides windows. 

The platform property in a nuget framework is also readonly, so we'll need a target platform moniker to use in initialization to set the platform property regardless.

sfoslund replied:
@dsplaisted per our discussion I've added the test case you suggested and made the code more explicit. Let me know if you have any other feedback.

davkean commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.TargetFrameworkInference.targets:
```diff
@@ -250,4 +250,19 @@ Copyright (c) .NET Foundation. All rights reserved.
+  <!--
+    Generate a list of valid TargetFramework aliases for the project system to consume for the target framework properties page drop down
+  -->
+  <Target Name="GenerateSupportedTargetFrameworkAlias">
```
I think this should say `Returns="@(SupportedTargetFrameworkAlias)`.

sfoslund replied:
I'm using the outputs attribute instead here, I think it's the same thing with different syntax: https://docs.microsoft.com/en-us/visualstudio/msbuild/target-element-msbuild?view=vs-2019#attributes

davkean replied:
The Outputs is on the _task_ not the target. We'll need the `Returns` or `Outputs` on the target to grab these values.

sfoslund committed:
PR feedback

sfoslund commented:
@cartermp @KathleenDollard @terrajobst As it's implemented right now, this would populate `net5.0-windows7.0` as the target framework when retargeting using the target framework drop down in VS. Do you see a problem with including this version number in the target framework?

sfoslund committed:
Reformatting SupportedTargetFrameworkAlias tests

cartermp commented:
Hmmm. I think that here's what it should be:

|App type|TFM string|
|--------|----------|
|WPF|`net5.0-windows`|
|WinForms|`net5.0-windows`|
|other windows apps|`net5.0-windows<version_no>`|
|iOS|`net5.0-ios<version_no>`|
|android|`net5.0-android<version_no>`|

Why no version suffix in WPF and Windows? Because it shouldn't matter because _all_ supported Windows versions will support all WPF and WinForms APIs.

sfoslund commented:
@cartermp your schema makes sense to me, I've implemented it that way and updated the tests to match. 

@dsplaisted I've adapted the tests as you suggested, this should be good for another review.

sfoslund committed:
Setting target platform to be versionless for windows forms and WPF

dsplaisted commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateSupportedTargetFrameworkAlias.cs:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        [WindowsOnlyTheory]
+        [InlineData("net5.0", "UseWpf")]
+        [InlineData("net5.0", "UseWindowsForms")]
```
This should be `net5.0-windows`, and I think there's a bug that will cause the resulting target framework to be `net5.0-windows-windows` or something in that case.

sfoslund committed:
PR feedback

terrajobst commented:
@sfoslund could you list the resulting display names for the UI?

sfoslund commented:
@terrajobst the display names will remain the same as before, so they're listed in [SupportedTargetFrameworks.props](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedTargetFrameworks.props)

terrajobst commented:
> @terrajobst the display names will remain the same as before, so they're listed in [SupportedTargetFrameworks.props](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.SupportedTargetFrameworks.props)

I meant the effect of [this code](https://github.com/dotnet/sdk/pull/13545/files#diff-11e7b7431abbd0596f030eefc10620e9R50-R51). Isn't that what is displayed in the UI?

sfoslund commented:
Lines 50-51 of GenerateSupportedTargetFrameworkAlias.cs are just copying the display name from the SupportedTargetFramework item into the new item with the correct target framework. The display name stays the same, this task just converts the target framework moniker into the target framework.

wli3 commented on src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToGenerateSupportedTargetFrameworkAlias.cs:
```diff
@@ -0,0 +1,156 @@
+        [InlineData("net5.0-windows", "UseWpf")]
+        [InlineData("net5.0-windows", "UseWindowsForms")]
+        [InlineData("netcoreapp3.1", "UseWpf")]
+        [InlineData("netcoreapp3.1", "UseWindowsForms")]
```
Is it ok to use task level tests? These tests are expensive (several seconds for tests that starts from a project on disk). And we add about 30 minutes every release, it is no negligible

sfoslund replied:
Sure, just converted to unit tests. Look okay to you (and @dsplaisted)?

dsplaisted replied:
We get a bit less coverage if we only do unit tests.  Mainly it doesn't cover how we set the default TargetPlatformVersion.

sfoslund replied:
We are able to test that the target platform version from the target platform moniker gets included in the target framework in [this test](https://github.com/dotnet/sdk/pull/13545/files/69c2724cdf27f4ecb484814744ca4ca17e9cedb2#diff-23f23314e105085486eb13f5b1650dccR54), is there more coverage you think we need?

sfoslund committed:
Converting GivenThatWeWantToGenerateSupportedTargetFrameworkAlias tests to task unit tests

vatsan-madhavan commented:
> Why no version suffix in WPF and Windows? Because it shouldn't matter because _all_ supported Windows versions will support all WPF and WinForms APIs.

This is not a good assumption/approach. 

Both WPF and WinForms have strong reasons to know what version of Win10 they are running on to have different fidelity of experience wrt DPI awareness. UWP targeting is not the only pivot that exists out there for platform targeting. 

/cc @merriemcgaw

merriemcgaw commented:
Agree with @vatsan-madhavan, we do often care about which version of Windows we're on, especially for DPI Awareness.

dsplaisted commented:
The design wasn't to say that Windows Forms or WPF, or apps written for them, can't care about the version of Windows they're running on.  Apps can still target a specific version of the OS if they want to, it just won't be represented explicitly in the TargetFramework by default.

The idea was that by default Windows Forms or WPF apps wouldn't need to target a specific version of Windows, and that Windows Forms and WPF themselves would do runtime lightup for OS versions.

vatsan-madhavan commented:
>and that Windows Forms and WPF themselves would do runtime lightup for OS versions.

This 'runtime lightup' business we do today is extremely complicated, and error-prone. I've fixed many bugs in this area, and also fixed bugs that have arisen because the OS added something to an enum that was unexpected that we didn't know how to handle and suddenly we have a high-volume crash in Visual Studio. 

If we are inventing a scheme that would transform 'runtime lightup' to 'versioned libraries', we should go that way in .NET 6 (w/o getting carried away and falling into a version-explosion trap) for WD libraries.

dsplaisted commented:
No, the general way to handle lightup over different OS versions is still runtime checks.  That seems to be the standard way of doing it for iOS, Android, and Windows.
